pkg/syncserver/sync_server.go

```diff
@@ -198,11 +251,85 @@ func (s *Server) Serve(cxt context.Context) {
 			encoder: gob.NewEncoder(conn),
 			readC:   make(chan interface{}),
 		}
-		go connection.handle()
+
+		go func() {
+			// Track the connection's lifetime in connIDToConn so we can kill it later if needed.
+			s.recordConnection(connection)
+			defer s.discardConnection(connection)
+			// Defer to the connection to do the handling.
+			connection.handle()
+		}()
+	}
+}
+
+func (s *Server) overConnLimit() bool {
+	s.connTrackingLock.Lock()
+	defer s.connTrackingLock.Unlock()
+	return len(s.connIDToConn) >= s.maxConns
+}
+
+func (s *Server) recordConnection(conn *connection) {
+	s.connTrackingLock.Lock()
+	s.connIDToConn[conn.ID] = conn
+	s.connTrackingLock.Unlock()
+}
+
+func (s *Server) discardConnection(conn *connection) {
+	s.connTrackingLock.Lock()
+	delete(s.connIDToConn, conn.ID)
+	s.connTrackingLock.Unlock()
+}
+
+func (s *Server) governNumberOfConnections(cxt context.Context) {
+	logCxt := log.WithField("thread", "numConnsGov")
+	maxConns := s.maxConns
+	ticker := jitter.NewTicker(s.dropInterval, s.dropInterval/10)
+	for {
+		select {
+		case newMax := <-s.maxConnsC:
+			if newMax == maxConns {
+				continue
+			}
+			s.connTrackingLock.Lock()
+			currentNum := len(s.connIDToConn)
+			s.connTrackingLock.Unlock()
+			logCxt.WithFields(log.Fields{
+				"oldMax":     maxConns,
+				"newMax":     newMax,
+				"currentNum": currentNum,
+			}).Info("New target number of connections")
+			maxConns = newMax
+			s.connTrackingLock.Lock()
+			s.maxConns = maxConns
+			s.connTrackingLock.Unlock()
+		case <-ticker.C:
+			s.connTrackingLock.Lock()
+			numConns := len(s.connIDToConn)
+			if numConns > maxConns {
+				for connID, conn := range s.connIDToConn {
+					logCxt.WithFields(log.Fields{
+						"max":     maxConns,
+						"current": numConns,
+						"connID":  connID,
+					}).Warn("Currently have too many connections, terminating one at random.")
+					conn.cancelCxt()
+					counterNumConnectionsDropped.Inc()
+					// Remove from the map now so that the connection count immediately drops;
+					// otherwise we might drop too many connections.
```

Comment:
That sounds worrying.  If there are other cases where we don't immediately remove from the map, perhaps there is a good reason for that.  On the other hand if there aren't any such cases, can we use common code to do immediate removal in all cases?